text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,723 @@
+you may assume the type of associated type defaults in provided methods.
+As noted in the [drawbacks] section,
+this would be useful for certain types of APIs.
+However, it is more likely than not that associated type defaults will
+be used as a mechanism for code reuse than for other constructs.
+As such, we consider the approach in this RFC to be the more ergonomic approach.
+
+Another alternative to the mechanism proposed in this RFC is to somehow
+track which methods rely on which associated types as well as constants.
+However, we have historically had a strong bias toward being explicit
+in signatures about such things, avoiding to infer them.
+With respect to semantic versioning, such an approach may also cause
+surprises for crate authors and their dependents alike.
+
+## Consistency with associated `const`s
+
+Consider the following valid example from stable Rust:
+
+```rust
+trait Foo {
+    const BAR: usize = 1;
+
+    fn baz() { println!("Hi I'm baz."); }
+}
+
+impl Foo for () {
+    fn baz() { println!("Hi I'm () baz."); }
+}
+```
+
+As we can see, you are permitted to override `baz` but leave `BAR` defaulted.
+This is consistent with the behaviour in this RFC in that it has the same
+property: *"you don't need to override all items if you override one"*.
+
+Consistency and uniformity of any programming language is vital to make
+its learning easy and to rid users of surprising corner cases and caveats.
+By staying consistent, as shown above, we can reduce the cost to our complexity
+budget that associated type defaults incur.
+
+## Overriding everything is less ergonomic
+
+We have already discussed this to some extent.
+Another point to consider is that Rust code frequently sports traits such as
+`Iterator` and `Future` that have many provided methods and few associated types.
+While these particular traits may not benefit from associated type defaults,
+many other traits, such as `Arbitrary` defined in the [motivation], would.
+
+## `default { .. }` is syntactically light-weight
+
+When you actually do need to assume the underlying default of an associated type
+in a provided method, `default { .. }` provides a syntax that is comparatively
+not *that* heavy weight.
+
+In addition, when you want to say that multiple items are overridable,
+`default { .. }` provides less repetition than specifying `default` on
+each item would. Thus, we believe the syntax is ergonomic.
+
+Finally, `default { .. }` works well and allows the user a good deal of control
+over what can and can't be assumed and what must be specialized together.
+The grouping mechanism also composes well as seen in
+[the section where it is discussed][default_groups].
+
+# Prior art
+[prior-art]: #prior-art
+
+## Haskell
+
+As Rust traits are a form of type classes,
+we naturally look for prior art from were they first were introduced.
+That language, being Haskell, permits a user to specify associated type defaults.
```

Comment:
While traits are type classes, it is also comparable to OOP interfaces; [specialization](https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md) in particular is entirely inspired from C++.

So I'd suggest adding the following comparison to prior arts as well:

* C++ (all diagnostic is done post-monomorphization so probably not an interesting example, but still...)
* Swift (Rust trait ≈ Swift protocols. Protocols can be inherited and can have associated types. But protocol constants and methods cannot be defaulted.)
* Perhaps also check what Scala is doing 🙃